File: AI Memory/memory_manager.py
"""
记忆管理器 - 对话存储、检索、上下文构建的核心模块。

数据存储在本地 JSON 文件中，结构如下：
{
    "version": "1.0",
    "conversations": [
        {
            "id": "uuid",
            "timestamp": "ISO8601",
            "topic": "对话主题摘要",
            "tags": ["tag1"],
            "messages": [
                {"role": "user", "content": "..."},
                {"role": "assistant", "content": "..."}
            ],
            "embedding": [0.1, 0.2, ...]
        }
    ]
}
"""

import json
import os
import uuid
from datetime import datetime, timezone
from typing import Optional

import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

from lightweight_embedding import LightweightEmbeddingService


class MemoryManager:
    """本地对话记忆管理器"""

    def __init__(
        self,
        storage_path: str = "memory.json",
        embedding_service: LightweightEmbeddingService = None,
    ):
        """
        Args:
            storage_path: JSON 存储文件路径
            embedding_service: 向量嵌入服务实例，不传则自动创建（默认 TF-IDF）
        """
        self.storage_path = storage_path
        self.embedding_service = embedding_service or LightweightEmbeddingService()
        self._data = self._load()

    # ========== 文件 I/O ==========

    def _load(self) -> dict:
        """从文件加载数据"""
        if os.path.exists(self.storage_path):
            with open(self.storage_path, "r", encoding="utf-8") as f:
                data = json.load(f)
            # 兼容旧格式：确保 conversations 是列表
            if "conversations" not in data:
                data["conversations"] = []
            return data
        return {"version": "1.0", "conversations": []}

    def _save(self):
        """保存数据到文件"""
        with open(self.storage_path, "w", encoding="utf-8") as f:
            json.dump(self._data, f, ensure_ascii=False, indent=2)

    # ========== CRUD ==========

    def add_conversation(
        self,
        messages: list[dict],
        topic: str = "",
        tags: list[str] = None,
        auto_embed: bool = True,
    ) -> str:
        """
        添加一条对话记录。

        Args:
            messages: [{"role": "user/assistant", "content": "..."}, ...]
            topic: 对话主题摘要，为空则自动用首条用户消息作为主题
            tags: 标签列表
            auto_embed: 是否自动生成嵌入向量

        Returns:
            对话 ID
        """
        conv_id = uuid.uuid4().hex[:12]

        # 自动生成主题
        if not topic:
            first_user_msg = next(
                (m["content"] for m in messages if m["role"] == "user"), ""
            )
            topic = first_user_msg[:80] + ("..." if len(first_user_msg) > 80 else "")

        conversation = {
            "id": conv_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "topic": topic,
            "tags": tags or [],
            "messages": messages,
            "embedding": None,
        }

        if auto_embed:
            conversation["embedding"] = self._compute_embedding_for_conv(conversation)

        self._data["conversations"].append(conversation)
        self._save()
        return conv_id

    def get_conversation(self, conv_id: str) -> Optional[dict]:
        """根据 ID 获取对话"""
        for conv in self._data["conversations"]:
            if conv["id"] == conv_id:
                return conv
        return None

    def delete_conversation(self, conv_id: str) -> bool:
        """删除对话，返回是否成功"""
        before = len(self._data["conversations"])
        self._data["conversations"] = [
            c for c in self._data["conversations"] if c["id"] != conv_id
        ]
        if len(self._data["conversations"]) < before:
            self._save()
            return True
        return False

    def list_conversations(self, limit: int = 20) -> list[dict]:
        """列出最近的对话（仅返回摘要信息，不含完整消息）"""
        convs = sorted(
            self._data["conversations"],
            key=lambda c: c["timestamp"],
            reverse=True,
        )
        return [
            {
                "id": c["id"],
                "timestamp": c["timestamp"],
                "topic": c["topic"],
                "tags": c["tags"],
                "message_count": len(c["messages"]),
            }
            for c in convs[:limit]
        ]

    # ========== 搜索 ==========

    def search(
        self,
        query: str,
        top_k: int = 5,
        min_score: float = 0.0,
        include_messages: bool = False,
    ) -> list[dict]:
        """
        语义搜索历史对话。

        Args:
            query: 搜索查询
            top_k: 返回条数
            min_score: 最低相似度阈值
            include_messages: 是否包含完整消息内容

        Returns:
            匹配的对话列表，按相似度降序
        """
        if not self._data["conversations"]:
            return []

        # 确保所有对话都有嵌入
        self._ensure_embeddings()

        # 提取所有嵌入向量
        embeddings = []
        valid_convs = []
        for conv in self._data["conversations"]:
            emb = conv.get("embedding")
            if emb is not None:
                embeddings.append(emb)
                valid_convs.append(conv)

        if not embeddings:
            return []

        # 计算相似度
        query_vec = self.embedding_service.encode_single(query).reshape(1, -1)
        emb_matrix = np.array(embeddings)
        similarities = cosine_similarity(query_vec, emb_matrix)[0]

        # 排序
        ranked = sorted(enumerate(similarities), key=lambda x: x[1], reverse=True)

        results = []
        for idx, score in ranked:
            if score < min_score:
                break
            conv = valid_convs[idx]
            result = {
                "id": conv["id"],
                "timestamp": conv["timestamp"],
                "topic": conv["topic"],
                "tags": conv["tags"],
                "score": float(score),
            }
            if include_messages:
                result["messages"] = conv["messages"]
            results.append(result)
            if len(results) >= top_k:
                break

        return results

    # ========== 上下文构建 ==========

    def build_context(
        self,
        query: str,
        top_k: int = 5,
        min_score: float = 0.3,
        max_context_chars: int = 8000,
    ) -> str:
        """
        为当前对话构建历史记忆上下文，可直接注入 LLM 的 system prompt。

        流程：语义搜索 → 按相似度排序 → 截断到 max_context_chars。

        Args:
            query: 当前用户问题
            top_k: 最多检索的对话数
            min_score: 最低相似度阈值
            max_context_chars: 上下文最大字符数（避免超出模型窗口）

        Returns:
            格式化的上下文字符串，无匹配时返回空字符串
        """
        results = self.search(query, top_k=top_k, min_score=min_score, include_messages=True)

        if not results:
            return ""

        lines = ["## 历史相关对话记录\n"]
        total_chars = len(lines[0])

        for r in results:
            # 格式化一条记录
            block = self._format_conversation_block(r)
            if total_chars + len(block) > max_context_chars:
                # 截断：只保留部分消息
                truncated = self._format_conversation_block(r, max_messages=2)
                if total_chars + len(truncated) <= max_context_chars:
                    lines.append(truncated)
                break
            lines.append(block)
            total_chars += len(block)

        return "\n".join(lines)

    def _format_conversation_block(self, result: dict, max_messages: int = None) -> str:
        """格式化单条对话记录"""
        timestamp = result["timestamp"][:19].replace("T", " ")
        lines = [
            f"### [{result['score']:.0%}] {result['topic']}",
            f"时间: {timestamp}",
            f"---",
        ]
        messages = result["messages"]
        if max_messages:
            messages = messages[:max_messages]

        for msg in messages:
            role_label = "用户" if msg["role"] == "user" else "助手"
            content = msg["content"]
            if len(content) > 500:
                content = content[:500] + "..."
            lines.append(f"**{role_label}**: {content}")

        lines.append("")  # 空行分隔
        return "\n".join(lines)

    # ========== 内部辅助 ==========

    def _compute_embedding_for_conv(self, conversation: dict) -> list:
        """为对话计算嵌入向量（使用 topic + 首条消息作为文本）"""
        text = conversation["topic"]
        # 加入更多上下文
        if conversation["messages"]:
            first_content = conversation["messages"][0]["content"][:500]
            text = text + " " + first_content
        vec = self.embedding_service.encode_single(text)
        return vec.tolist()

    def _ensure_embeddings(self):
        """确保所有对话都有嵌入向量（兼容增量更新）"""
        needs_save = False
        for conv in self._data["conversations"]:
            if conv.get("embedding") is None:
                conv["embedding"] = self._compute_embedding_for_conv(conv)
                needs_save = True
        if needs_save:
            self._save()

    def rebuild_embeddings(self):
        """重建所有对话的嵌入向量（切换模型后使用）"""
        for conv in self._data["conversations"]:
            conv["embedding"] = self._compute_embedding_for_conv(conv)
        self._save()
        print(f"已重建 {len(self._data['conversations'])} 条对话的嵌入向量")

    @property
    def count(self) -> int:
        """总对话数"""
        return len(self._data["conversations"])

File: electron/services/memory/memory-manager.ts
/**
 * 记忆管理器 — 对话存储、语义检索、上下文构建
 *
 * 对标 Python 版 MemoryManager，纯 TypeScript 实现。
 * 数据存储在本地 JSON 文件中。
 */

import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { TfidfVectorizer } from './tfidf-vectorizer';
import type {
  MemoryConversation,
  MemoryMessage,
  ConversationSummary,
  SearchResult,
  MemoryData,
  MemoryStats,
} from './types';

export class MemoryManager {
  private storagePath: string;
  private indexPath: string;
  private vectorizer: TfidfVectorizer;
  private data: MemoryData;

  constructor(storagePath: string) {
    this.storagePath = storagePath;
    this.indexPath = storagePath.replace(/\.json$/, '-index.json');
    this.vectorizer = new TfidfVectorizer(this.indexPath);
    this.data = this.load();
  }

  // ==================== 文件 I/O ====================

  private load(): MemoryData {
    try {
      if (fs.existsSync(this.storagePath)) {
        const raw = fs.readFileSync(this.storagePath, 'utf-8');
        const data = JSON.parse(raw) as MemoryData;
        if (!data.conversations) data.conversations = [];
        return data;
      }
    } catch {
      // 文件损坏，重新开始
      console.warn('[MemoryManager] 记忆文件损坏，将重新创建');
    }
    return { version: '1.0', conversations: [] };
  }

  private save(): void {
    const dir = path.dirname(this.storagePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    fs.writeFileSync(this.storagePath, JSON.stringify(this.data, null, 2), 'utf-8');
  }

  // ==================== CRUD ====================

  /**
   * 添加一条对话到记忆库
   * @returns 记忆条目 ID
   */
  addConversation(
    conversationId: string,
    title: string,
    messages: MemoryMessage[],
    tags: string[] = [],
  ): string {
    const id = crypto.randomUUID();

    const conv: MemoryConversation = {
      id,
      conversationId,
      title: title || this.extractTopic(messages),
      timestamp: Date.now(),
      tags,
      messages,
      embedding: null,
    };

    // 生成嵌入向量
    conv.embedding = this.computeEmbedding(conv);

    this.data.conversations.push(conv);
    this.save();
    return id;
  }

  /**
   * 更新已有对话（或追加新消息）
   */
  updateConversation(id: string, messages: MemoryMessage[], title?: string): boolean {
    const idx = this.data.conversations.findIndex((c) => c.id === id);
    if (idx === -1) return false;

    this.data.conversations[idx].messages = messages;
    if (title) this.data.conversations[idx].title = title;
    this.data.conversations[idx].timestamp = Date.now();
    this.data.conversations[idx].embedding = this.computeEmbedding(this.data.conversations[idx]);

    this.save();
    return true;
  }

  /** 删除对话 */
  deleteConversation(id: string): boolean {
    const before = this.data.conversations.length;
    this.data.conversations = this.data.conversations.filter((c) => c.id !== id);
    if (this.data.conversations.length < before) {
      this.save();
      return true;
    }
    return false;
  }

  /** 根据记忆 ID 获取对话 */
  getConversation(id: string): MemoryConversation | null {
    return this.data.conversations.find((c) => c.id === id) || null;
  }

  /** 根据 ChatAI 对话 ID 查找记忆 */
  findByConversationId(conversationId: string): MemoryConversation | null {
    return this.data.conversations.find((c) => c.conversationId === conversationId) || null;
  }

  /**
   * 插入或更新记忆（按 conversationId 匹配）
   * @returns 记忆条目 ID
   */
  upsertByConversationId(
    conversationId: string,
    title: string,
    messages: MemoryMessage[],
    tags: string[] = [],
  ): string {
    const existing = this.findByConversationId(conversationId);
    if (existing) {
      this.updateConversation(existing.id, messages, title);
      return existing.id;
    }
    return this.addConversation(conversationId, title, messages, tags);
  }

  /**
   * 列出所有对话摘要（不含完整消息）
   */
  listConversations(limit: number = 50): ConversationSummary[] {
    const sorted = [...this.data.conversations].sort((a, b) => b.timestamp - a.timestamp);
    return sorted.slice(0, limit).map((c) => ({
      id: c.id,
      conversationId: c.conversationId,
      timestamp: c.timestamp,
      topic: c.title,
      tags: c.tags,
      messageCount: c.messages.length,
    }));
  }

  // ==================== 搜索 ====================

  /**
   * 语义搜索历史对话
   */
  search(
    query: string,
    topK: number = 5,
    minScore: number = 0,
    includeMessages: boolean = false,
  ): SearchResult[] {
    if (this.data.conversations.length === 0) return [];

    // 确保所有对话都有嵌入向量
    this.ensureEmbeddings();

    // 重建向量化器索引
    this.rebuildVectorizer();

    const queryVec = this.vectorizer.transform(query);
    if (queryVec.length === 0) return [];

    // 计算余弦相似度
    const scored = this.data.conversations
      .map((conv) => {
        if (!conv.embedding || conv.embedding.length === 0) return { conv, score: 0 };
        const score = this.cosineSimilarity(queryVec, conv.embedding);
        return { conv, score };
      })
      .filter((r) => r.score >= minScore)
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);

    return scored.map((r) => ({
      id: r.conv.id,
      conversationId: r.conv.conversationId,
      timestamp: r.conv.timestamp,
      topic: r.conv.title,
      tags: r.conv.tags,
      score: Math.round(r.score * 10000) / 10000,
      messages: includeMessages ? r.conv.messages : undefined,
    }));
  }

  // ==================== 上下文构建 ====================

  /**
   * 为当前对话构建历史记忆上下文
   * 返回可直接注入 LLM system prompt 的文本
   */
  buildContext(
    query: string,
    topK: number = 5,
    minScore: number = 0.2,
    maxContextChars: number = 8000,
  ): string {
    const results = this.search(query, topK, minScore, true);
    if (results.length === 0) return '';

    const lines = ['## 历史相关对话记录'];
    let totalChars = lines[0].length;

    for (const r of results) {
      const block = this.formatContextBlock(r);
      if (totalChars + block.length > maxContextChars) {
        // 尝试截断版本
        const truncated = this.formatContextBlock(r, 2);
        if (totalChars + truncated.length <= maxContextChars) {
          lines.push(truncated);
        }
        break;
      }
      lines.push(block);
      totalChars += block.length;
    }

    return lines.join('\n');
  }

  /**
   * 格式化单条对话为上下文字符串
   */
  private formatContextBlock(result: SearchResult, maxMessages?: number): string {
    const date = new Date(result.timestamp).toISOString().slice(0, 19).replace('T', ' ');
    const scorePercent = Math.round(result.score * 100);
    const lines: string[] = [
      `### [相关度 ${scorePercent}%] ${result.topic}`,
      `时间: ${date}`,
      '---',
    ];

    let messages = result.messages || [];
    if (maxMessages) messages = messages.slice(0, maxMessages);

    for (const msg of messages) {
      const roleLabel = msg.role === 'user' ? '用户' : '助手';
      let content = msg.content;
      if (content.length > 500) content = content.slice(0, 500) + '...';
      lines.push(`**${roleLabel}**: ${content}`);
    }

    lines.push('');
    return lines.join('\n');
  }

  // ==================== 辅助方法 ====================

  /** 统计信息 */
  getStats(): MemoryStats {
    const totalMessages = this.data.conversations.reduce(
      (sum, c) => sum + c.messages.length,
      0,
    );
    return {
      totalConversations: this.data.conversations.length,
      totalMessages,
      lastUpdated: this.data.conversations.length > 0
        ? Math.max(...this.data.conversations.map((c) => c.timestamp))
        : 0,
    };
  }

  /** 重建所有嵌入向量（切换模型或修复数据时使用） */
  rebuildEmbeddings(): void {
    for (const conv of this.data.conversations) {
      conv.embedding = this.computeEmbedding(conv);
    }
    this.save();
    this.rebuildVectorizer();
  }

  /** 清空全部记忆 */
  clearAll(): void {
    this.data.conversations = [];
    this.save();
  }

  // ==================== 内部方法 ====================

  /** 为对话计算嵌入向量 */
  private computeEmbedding(conv: MemoryConversation): number[] {
    // 使用标题 + 首条用户消息的前500字符作为源文本
    let text = conv.title;
    const firstUserMsg = conv.messages.find((m) => m.role === 'user');
    if (firstUserMsg) {
      text += ' ' + firstUserMsg.content.slice(0, 500);
    }
    return this.vectorizer.transform(text);
  }

  /** 确保所有对话都有嵌入向量 */
  private ensureEmbeddings(): void {
    let needsSave = false;
    for (const conv of this.data.conversations) {
      if (!conv.embedding || conv.embedding.length === 0) {
        conv.embedding = this.computeEmbedding(conv);
        needsSave = true;
      }
    }
    if (needsSave) this.save();
  }

  /** 用所有对话重新拟合 TF-IDF 向量化器 */
  private rebuildVectorizer(): void {
    const documents = this.data.conversations.map((c) => {
      let text = c.title;
      const firstUserMsg = c.messages.find((m) => m.role === 'user');
      if (firstUserMsg) text += ' ' + firstUserMsg.content.slice(0, 500);
      return text;
    });
    if (documents.length > 0) {
      this.vectorizer.fitTransform(documents);
      // 用新拟合的向量化器重新计算嵌入
      for (let i = 0; i < this.data.conversations.length; i++) {
        this.data.conversations[i].embedding = this.vectorizer.transform(documents[i]);
      }
    }
  }

  /** 从消息中提取主题 */
  private extractTopic(messages: MemoryMessage[]): string {
    const firstUserMsg = messages.find((m) => m.role === 'user');
    if (!firstUserMsg) return '未命名对话';
    const content = firstUserMsg.content.trim();
    return content.length > 80 ? content.slice(0, 80) + '...' : content;
  }

  /** 余弦相似度 */
  private cosineSimilarity(a: number[], b: number[]): number {
    if (a.length === 0 || b.length === 0) return 0;
    const len = Math.min(a.length, b.length);
    let dot = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < len; i++) {
      dot += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }
    const denom = Math.sqrt(normA) * Math.sqrt(normB);
    return denom === 0 ? 0 : dot / denom;
  }
}
